core/CronArchive.php

```diff
@@ -399,159 +315,100 @@ public function run()
-                        }
-                    } catch (UnexpectedWebsiteFoundException $e) {
-                        $this->logger->info("Skipped website id $idSite, got: UnexpectedWebsiteFoundException");
-                        continue;
-                    }
-
-                } elseif ($hasWebsiteDayFinishedSinceLastRun) {
-                    $this->logger->info("Day has finished for website id $idSite since last run");
-                } elseif ($isOldReportInvalidatedForWebsite) {
-                    $this->logger->info("Old report was invalidated for website id $idSite");
+            // get archives to process simultaneously
+            $archivesToProcess = [];
+            while (count($archivesToProcess) < $countOfProcesses) {
+                $invalidatedArchive = $this->getNextInvalidatedArchive($periodToCheckFor = null);
+                if (empty($invalidatedArchive)) {
+                    $this->logger->debug("No next invalidated archive.");
+                    break;
                 }
-            }
 
-            /**
-             * This event is triggered before the cron archiving process starts archiving data for a single
-             * site.
-             *
-             * @param int $idSite The ID of the site we're archiving data for.
-             */
-            Piwik::postEvent('CronArchive.archiveSingleSite.start', array($idSite));
+                if ($this->isDoneFlagForPlugin($invalidatedArchive['name'])) {
+                    $this->logger->debug("Found plugin specific invalidated archive, ignoring.");
+                    $this->addIdArchivesToExclude($invalidatedArchive);
+                    continue;
+                }
 
-            $completed = $this->archiveSingleSite($idSite, $requestsBefore);
+                if ($this->archiveArrayContainsArchive($archivesToProcess, $invalidatedArchive)) {
+                    $this->logger->debug("Found duplicate invalidated archive {$invalidatedArchive['idarchive']}, ignoring.");
+                    $this->addIdArchivesToExclude($invalidatedArchive);
+                    continue;
+                }
 
-            /**
-             * This event is triggered immediately after the cron archiving process starts archiving data for a single
-             * site.
-             *
-             * @param int $idSite The ID of the site we're archiving data for.
-             */
-            Piwik::postEvent('CronArchive.archiveSingleSite.finish', array($idSite, $completed));
-        } while (!empty($idSite));
+                $reason = $this->shouldSkipArchive($invalidatedArchive);
+                if ($reason) {
+                    $this->logger->debug("Skipping invalidated archive {$invalidatedArchive['idarchive']}: $reason");
+                    $this->addIdArchivesToExclude($invalidatedArchive);
+                    continue;
+                }
 
-        $this->logger->info("Done archiving!");
+                $idArchive = $this->model->startArchive(
```

Comment:
> do I see this right that we are now launching an archiver for each specific day that needs to be archived instead of say last3 we would launch 3 archive commands for each specific day?

Yes, it will launch a command for each day. The archives for each of these days will definitely have visits, though, since there's an optimization to skip in that case. So 100ms shouldn't be an issue if archiving for a day takes 10s or more?

> I suppose if a new segment is added, then it could launch a lot of commands compared to before?

Yes, it will definitely run a lot of commands in this case... If say 3000 days are invalidated, that's 300s added to the total execution (if startup is 100ms). Would this be an issue in most cases?

It would be nice to be able to test w/ a big instance to see how well it works in certain cases (eg, no invalidations, invalidating some dates in the past, new segment invalidating 3 years in the past).

Reply:
I reckon startup is somewhat around 100-200ms. So if it adds 6-12 minutes that might be fine as it would be only eg if adding a new segment, or changing it, or ... I reckon we'll only be really able to see the impact after some time but could also test it. I'd say once we have a first beta released we could test it eg on demo and run it there for a while to see what it does?